Integration Docs: Tailwind (#3429)
withastro/astro: packages/integrations/tailwind/README.md

# @astrojs/tailwind 💨

This **[Astro integration][astro-integration]** brings [Tailwind's](https://tailwindcss.com/) utility CSS classes to every `.astro` file and [framework component](https://docs.astro.build/en/core-concepts/framework-components/) in your project, along with support for the Tailwind configuration file.

https://user-images.githubusercontent.com/4033662/169918388-8ed153b2-0ba0-4b24-b861-d6e1cc800b6c.mp4

## Why Tailwind?

Tailwind lets you use utility classes instead of writing CSS. These utility classes are mostly one-to-one with a certain CSS property setting: for example, adding the `text-lg` to an element is equivalent to setting `font-size: 1.125rem` in CSS. You might find it easier to write and maintain your styles using these predefined utility classes!

> If you don't like those predefined settings, you can [customize the Tailwind configuration file](https://tailwindcss.com/docs/configuration) to your project's design requirements. For exmaple, if the "large text" in your design is actually `2rem`, you can [change the `lg` fontSize setting](https://tailwindcss.com/docs/font-size#customizing-your-theme) to `2rem`.

Reviewer: ```suggestion
If you don't like those predefined settings, you can [customize the Tailwind configuration file](https://tailwindcss.com/docs/configuration) to your project's design requirements. For example, if the "large text" in your design is actually `2rem`, you can [change the `lg` fontSize setting](https://tailwindcss.com/docs/font-size#customizing-your-theme) to `2rem`.
```

> Tailwind is also a great choice to add styles to a React, Preact, or Solid component, which don't support a `<style>` tag in the component file. That said, it's generally discouraged to use both Tailwind and another styling method in the same file.

Author: That's what I meant. I think it's important advice that Tailwind isn't usually used with another styling solution, but we can revise

- <strong>[Why Tailwind](#why-tailwind)</strong>
- <strong>[Installation](#installation)</strong>
- <strong>[Usage](#usage)</strong>
- <strong>[Configuration](#configuration)</strong>
- <strong>[Examples](#examples)</strong>
- <strong>[Troubleshooting](#troubleshooting)</strong>
- <strong>[Contributing](#contributing)</strong>
- <strong>[Changelog](#changelog)</strong>

## Why Tailwind?

Tailwind lets you use utility classes instead of writing CSS. These utility classes are mostly one-to-one with a certain CSS property setting: for example, adding the `text-lg` to an element is equivalent to setting `font-size: 1.125rem` in CSS. You might find it easier to write and maintain your styles using these predefined utility classes!

If you don't like those predefined settings, you can [customize the Tailwind configuration file](https://tailwindcss.com/docs/configuration) to your project's design requirements. For example, if the "large text" in your design is actually `2rem`, you can [change the `lg` fontSize setting](https://tailwindcss.com/docs/font-size#customizing-your-theme) to `2rem`.

> Tailwind is also a great choice to add styles to React, Preact, or Solid components, which don't support a `<style>` tag in the component file. That said, it's generally discouraged to use both Tailwind and another styling method in the same file.

Reviewer: ```suggestion
Tailwind is also a great choice to add styles to React, Preact, or Solid components, which don't support a `<style>` tag in the component file. 

Note: it's generally discouraged to use both Tailwind and another styling method (e.g. Styled Components) in the same file.
```

Author: Thanks!

  integrations: [tailwind()],
}
```
  
Then, restart the dev server.
</details>

> ## Usage

Reviewer: Visually, this hierarchy feels a little off.

The big heading is Usage, but all the subheadings are configuration-related. And, when I see "usage" I would probably expect to see code samples, Tailwind actually being used in an Astro component. What about a screenshot here, for the people who are skimming headers and didn't watch your video? "Oh, here's where I can jump to to see how to USE Tailwind in Astro..." and then I feel like this doesn't give them that.  :)

Does this also point to maybe Configuration deserves its own heading? 

![Screenshot 2022-05-24 10 50 49](https://user-images.githubusercontent.com/5098874/170052020-23e2b156-d8d9-4033-8557-ca37f746efa9.png)

Author: Configuration is part of usage. In order to use something the way you want it, you must configure it

Author: For other integrations I absolutely agree - this is the spot for screenshots! But for this one, I feel like I got there already in the intro, and the most important point I want to make is GO TO TAILWIND DOCS 😆

### Configuring Tailwind

If you used the Quick Install instructions and said yes to each prompt, you'll see a `tailwind.config.cjs` file in your project's root directory. Use this file for your Tailwind configuration changes. You can learn how to customize Tailwind using this file [in the Tailwind docs](https://tailwindcss.com/docs/configuration).

If it isn't there, you add your own `tailwind.config.(js|cjs|mjs)` file to the root directory and the integration will use its configurations. This can be great if you already have Tailwind configured in aother project and want to bring those settings over to this one.

### Configuring the Integration

The Astro Tailwind integration handles the communication between Astro and Tailwind and it has its own options. Change these in the `astro.config.mjs` file (_not_ the Tailwind configuration file) which is where your project's integration settings live.

<details>

>   <summary>config.path</summary>

Reviewer 2: We can maintain the heading hierarchy/styling by setting a heading here I think. Does introduce some margin, so if that seems visually unappealing, maybe just `<strong>`?

```suggestion
  <summary><h4>config.path</h4></summary>
```

Reviewer: If some of these concerns are about how this will look on npmjs.com, should we maybe merge and be our own guinea pig?  😅 

I can spend the today with this, and make quick changes as necessary. @Jutanium @delucis

  })],
}
```
</details>

> ## Examples

Author: Hmm maybe we can do that with a table? not sure

- The [Astro Tailwind Starter](https://github.com/withastro/astro/tree/latest/examples/with-tailwindcss?on=github) gets you up and running with a base for your project that uses Tailwind for styling
- Astro's homepage uses Tailwind. Check out its [Tailwind configuration file](https://github.com/withastro/astro.build/blob/main/tailwind.config.js) or an [example component](https://github.com/withastro/astro.build/blob/main/src/components/integrations/IntegrationCard.astro)
- The [Astro Ink](https://github.com/one-aalam/astro-ink), [Sarissa Blog](https://github.com/iozcelik/SarissaBlogAstroStarter), and [Creek](https://github.com/robertguss/Astro-Theme-Creek) themes use Tailwind for styling
- [Browse Astro Tailwind projects on GitHub](https://github.com/search?q=%22%40astrojs%2Ftailwind%22+filename%3Apackage.json&type=Code) for more examples!

## Troubleshooting
- If your installation doesn't seem to be working, make sure to restart the dev server.
- If you edit and save a file and don't see your site update accordingly, try refreshing the page.

> - Occasionally, Astro's dev server will run into a problem, and refreshing the page still won't work. In these cases, restart the dev server.

Reviewer: I feel like *maybe* we don't want to hit this?  😅 

```suggestion
- If you edit and save a file and don't see your site update accordingly, try refreshing the page.
- If refreshing the page doesn't update your preview, or if a new installation doesn't seem to be working, then restart the dev server.
```
Maybe sneak it in something like that?

> For help, check out the `#support-threads` channel on Discord. Our friendly Support Squad members are here to help!

Reviewer: ```suggestion
For help, check out the `#support-threads` channel on [Discord](https://astro.build/chat). Our friendly Support Squad members are here to help!
```

You can also check our [Astro Integration Documentation][astro-integration] for more on integrations.

[astro-integration]: https://docs.astro.build/en/guides/integrations-guide/
[astro-ui-frameworks]: https://docs.astro.build/en/core-concepts/framework-components/#using-framework-components

## Contributing

> This package is maintained by Astro's Core team. You're welcome to submit an issue or PR!

Reviewer: Depending on how much we want to encourage this, we could be linking to "new issue", "new PR" here. I know the reader is RIGHT HERE, so not terribly important. But I think a link demonstrates being open to contributions. Just depends if it fits the vibe... and if that's what we in fact want to encourage.

Author: This points to a broader problem of the New Issue button being on the _entire_ Astro repo rather than this project... is there a way to link to creating a new issue with "Tailwind Integration:" already filled out in the header?

Reviewer: Agreed.  So I checked, and you can enable (and seems like withastro/astro *has* enabled) the You must choose a type of issue screen.

The problem is it looks like this, and only "bug report" actually files an issue in this repo. If you choose documentation, it's a link to the docs repo. 

![Screenshot 2022-05-27 07 44 45](https://user-images.githubusercontent.com/5098874/170685220-1e6fbc06-cfa6-47cd-a04f-a08eaec787ef.png)

The most-proper but I think least-sensible option is to have another choice with bug report that is "correct integration documentation." But that is objectively terrible, especially given that to make issues with REGULAR documentation is already a choice and takes you somewhere different.

So, one imperfect option is to link directly to Docs issues where we will triage, and the workflow becomes to require an issue to Docs to tell us there's something wrong, where we can then guide them through the process if they want to actually submit a PR for it in the proper place. (The default option being that we make the fix ourselves, after the issue alerts us to any errors.)
